### PR TITLE
Several fixes to the ALVA driver

### LINKED ISSUE #8074: Alva satellite used via protocol converter and Optelec BC6 driver does not detect some keys and some keys were assigned twice
First of all I would like to thank all of you to bring support for protocol converter into NVDA.

Here is one problem: Alva Satellite has 2 Satellite keypads, one on the leftside of display and one on the rightside with 6 keys on the each keypads, but NVDA treats them as one keypad. I do not know if it is bug of NVDA or bug of protocol converter, so I have opened this ticket as @leonardder suggested in private conversation.

### Steps to reproduce:

1. connect Alva Satellite to computer via protocol converter
2. wait system to install it on the system if it was not installed already
3. from the braille settings choose Optelec bc6...
4. turn input help on
5. play with keys on Satellite display

### Expected behavior:

NVDA should differentiate between Satellite keypads, so NVDA could emulates most of keys which are on bc6 displays.

### Actual behavior:

Now NVDA emulates: sp1, sp4, and sp left, right, up and down on both satellite keypads. So there are not: sp2, sp3, spEnter, etouch1, etouch3 and etouch4.

### System configuration:
NVDA version:

next: 14941,891a9c9e

NVDA Installed or portable:
installed

Other information:

Windows version:

Windows 10Ver1709 (64-bit) build 16299.248

### Other questions:

Does the issue still occur after restarting your PC?
yes

Have you tried any other versions of NVDA?
no

### LINKED ISSUE #8106: Alva BC680 doesn't initialize or initializes with wrong number of cells
On Sunday, 18th March 2018, I sent an e-mail to info@optelec.de and as CC to 'info@nvaccess.org' From Optelec GmbH I got the info that the latest Firmware for the Optelec ALVA BC680 is 3.0.2 and that the main developer in the Netherlands has nothing to do with the new ALVA-driver in NVDA 2018.1.

As I have a few other time-consuming task to do in the upcoming days and weeks I just use google Translate for translating my German e-mail into English. If something makes no sense, just let me know and I will translate it manually. You will find the German original below the English translation.

The "nvda.ini" in the user-config-folder has the same checksum (SHA-256) as the one in the systemconfig-folder in (C:\Programme (x86)\NVDA). And in addition to #8 the ALVA BC680 also hangs immediately after connecting and disconnecting a Bluetooth connection to a iPhone (iOS 10.3.3)or tablet (Fire OS 5.6.0.1 with BrailleBack). This issue exists since NVDA 2018.1 while the braille display is connected via USB2.

===== Translation of the German e-mail into English by Google Translate =====
[...] Firmware version of my Optelec ALVA BC680: 3.0.2 (Bluetooth version: 1.20.0)
There is no network and no keyboard (USB-A socket on the Braille display) connected to the ALVA BC680. The USB1 port (left) is connected to my stand PC with my NUC and the USB2 interface (right).
In NVDA 2017.4, the Braille display driver via USB1 also worked perfectly smoothly via USB2. In NVDA 2018.1, the driver was rewritten from scratch, and I also had to reiterate the importance of the functions of the Braille display. But I would not write an e-mail now, if that had already become everything. There are a lot of new bugs in NVDA 2018.1.
I would like to ask you to e-mail the firmware version for the Optelec ALVA BC680. Maybe you can fix a bug here - maybe.
My main system (stand-pc): Windows 7 64-bit SP1, NVDA 2018.1 is installed, update of the current version of the BC680 via USB2 with the PC (USB 2.0 port directly on the mainboard)
Note: The ALVA BC680 is installed via USB1 with the NUC (Intel NUC5i3RYH, Win7x64Sp1, NVDA 2015.3, here no NVDA update planned for 2018.1 in the near future), connected to USB2 to the USB 3.0 port does not work properly because the Intel USB drivers under Windows 7 do not work properly. However, the connection via USB1 (possibly in the Braille itself) also works perfectly.

Here is the new list of bugs that do not yet apply with NVDA 2017.4:
1. After the boot process on the logon screen, only the first characters on the Braille display are displayed via USB2. NVDA does not recognize that the Braille display has 80 modules. In the system menu of the Braille display, the braille display division is disabled, especially since the default value of "5 characters" has been changed to "30" characters "for testing purposes." What this value of "5 characters" comes from at once is beyond me Braille output can be problematic to move these five characters forward and backward, and if the braille display is connected via USB1, this will not cause these errors.
2. Immediately after logging in, here too NVDA outputs only the five characters on the Braille display when connected via USB2. With USB1, this error does not occur.
3. If NVDA is restarted via the desktop connection, the Braille display via USB2 will not be known. The "Braille Settings" dialog box says "No braille display". If NVDA is started again via this connection, this will be announced again. Also this problem occurs only on USB2, but not on USB1.
4. If you open the dialog window "Braille Settings", "Braille display: Optelec ALVA 6-Series / Protocol converter" is selected, and closes with "OK", the following error message appears: "Error while driving the Braille display! ALVA "could not be changed." If you close the window again with "OK" (or the Enter key), the ALVA BC680 will be correctly recognized again via USB2. But it can also be that here too more than ever 80 characters are displayed. follow together: "double" is poached, it goes again.
5. In the "protected desktop" (CTRL + ALT + Del) there will be no more output on the Braille display, even if it has been knocked down at the user level. That was not the case with NVDA 2017.4. Here, the Braille content was automatically switched in both environments. In NVDA 2018.1, "Protected Desktop" was on the Braille display, so CTRL + ALT + Del was pressed and the braille display was paired at the user level.
6. Under NVDA 2017.4, when the PC was put into standby mode, a blinking cursor (points 7 and 8) was briefly shown on the ALVA BC680 due to the password input text box. This is no longer the case with NVDA 2018.1. (Is related to point 5)
7. If you switch on the Braille display first and wake up the PC from the standby mode only after that, the cursor blinked under NVDA 2017.4 after a few seconds in the password input text field again. This is no longer the case with NVDA 2018.1. (Is related to point 5)
8. If you switch the connection types between USB (which means USB1), USB2 and BT (that means Bluetooth) with the right smartpad buttons SP2 + SP3 + SPUp or SP2 + SP3 + SPDown, the braille display responds after the switch back to USB2 under NVDA 2018.1 no longer. No new text is output on the Braille display, nor do the buttons on the Braille display work. The braille display has to be "double" reconnected so that it becomes usable again. It also does not matter whether a connection via USB1 or via Bluetooth to another device has actually been made before switching back to "USB2". Via USB1 this error does not occur under NVDA 2018.1. Under NVDA 2017.4, this error does not occur on either USB1 or USB2.
9. By simultaneously pressing SP1 + SP2 + SP3, SP2 + SP3 + Sp4 and SP1 + SP2 + SP3 + SP4, at least one of these four smartpad buttons will "hang" under both NVDA 2018.1 via USB1 and via USB2. If you then immediately press, for example "Etouch1", NVDA recognizes "SP1 + Sp2 + Sp3 + etouch1". By repeatedly pressing each of the four aforementioned smartpad buttons, this "snagging" can sometimes be released. Alternatively, the braille display needs to be "double" reconnected so that all control buttons on the ALVA BC680 are working properly again. Under NVDA 2017.4 this problem did not occur either.
10. [Describes another bug that already exists in NVDA 2015.3. That's why he does not stand here.]

===== Original e-mail in German =====
[…] Firmware-Version meiner Optelec ALVA BC680: 3.0.2 (Bluetooth-Version: 1.20.0)
Es ist kein Netzteil und auch keine Tastatur (USB-A-Buchse links an der Braillezeile) an der ALVA BC680 angeschlossen. Der USB1-Port (links) ist mit meinem NUC und der USB2-Port (rechts) mit meinem Stand-PC verbunden.

In NVDA 2017.4 funktionierte der Braillezeilen-Treiber via USB1 sowie via USB2 absolut problemlos. In NVDA 2018.1 wurde jener Treiber von Grund auf neu geschrieben, weshalb ich auch sämtliche Bedientasten der Braillezeile wieder den von mir gewünschten Funktionen zuweisen musste. Ich würde Ihnen jetzt aber keine E-Mail extra schreiben, wenn das schon alles gewesen wäre. Es kamen nämlich einige zahlreiche neue Bugs in NVDA 2018.1 dazu.

Zuvor darf ich Sie allerdings darum bitten mir so rasch wie nur möglich die aktuellste Firmware-Version für die Optelec ALVA BC680 via E-Mail zukommen zu lassen. Vielleicht können hierdurch ein paar der neuen Bugs gleich behoben werden – vielleicht.

Mein Haupt-System (Stand-PC): Windows 7 64-Bit SP1, NVDA 2018.1 ist installiert (keine portable Kopie, Update erfolgte heute Morgen von 2017.4 auf 2018.1), Anschluss der ALVA BC680 via USB2 mit dem PC (USB-2.0-Port direkt am Motherboard)
Anmerkung am Rande: Die ALVA BC680 ist via USB1 mit dem NUC (Intel NUC5i3RYH, Win7x64Sp1, NVDA 2015.3 installiert, hier kein NVDA-Update auf 2018.1 in der nahen Zukunft geplant) angeschlossen, da dort der Anschluss via USB2 an den USB-3.0-Port nicht korrekt funktioniert, weil die Intel-USB-Treiber unter Windows 7 nicht vollumfänglich ordentlich arbeiten. Der Anschluss via USB1 (möglicherweise dank dem USB-Hub in der Braillezeile selber) funktioniert aber tadellos.

Hier nun die Auflistung der neuen Bugs, die mit NVDA 2017.4 noch nicht auftragen:
1.	Nach dem Boot-Vorgang auf dem Logon-Screen werden via USB2 nur die ersten fünf Zeichen ganz links auf der Braillezeile angezeigt. NVDA erkennt hier nicht, dass die Braillezeile über 80 Module verfügt. Im System-Menü der Braillezeile ist die Braillezeilen-Splittung deaktiviert, zumal ich dort sowieso den Standardwert von „5 Zeichen“ auf „30“ Zeichen“ zu Testzwecken geändert gehabt hatte. Woher folglich dieser Wert von „5 Zeichen“ auf einmal herkommen soll, ist mir schleierhaft. Die Braillezeilen-ausgabe kann allerdings problemlos um diese fünf Zeichen nach vor und zurück gescrollt werden. Ist die Braillezeile via USB1 verbunden, so tritt dieser Fehler nicht auf.
2.	Unmittelbar nach der Anmeldung gibt auch hier NVDA nur die fünf Zeichen auf der Braillezeile aus, wenn jene via USB2 angeschlossen ist. Bei USB1 tritt dieser Fehler nicht auf.
3.	Wird NVDA über die Desktop-Verknüpfung neu gestartet, so wird die Braillezeile via USB2 nicht erkannt. Im Dialogfenster „Braille-Einstellungen“ steht „Keine Braillezeile“. Wird NVDA ein weiteres Mal über dieselbe Verknüpfung neu gestartet, so wird jene wieder korrekt erkannt. Auch dieses Problem tritt nur via USB2, nicht aber via USB1 auf.
4.	Öffnet man das Dialogfenster „Braille-Einstellungen“, wobei „Braillezeile: Optelec ALVA 6-Serien/Protokoll-Konverter“ bereits ausgewählt ist, und schließt jenes mit „OK“, so erscheint folgende Fehlermeldung: „Fehler beim Ansteuern der Braillezeile! Die Braillezeile "ALVA" konnte nicht angesteuert werden.“ Schließt man dieses Dialogfenster nun gleich wieder mit „OK“ (oder die Enter-Taste), so wird die ALVA BC680 via USB2 meistens wieder korrekt erkannt. Es kann aber auch sein, dass auch hier plötzlich dann mur mehr fünf statt 80 Zeichen angezeigt werden. folglich muss wieder „doppelt“ gekoppelt werden, damit es wieder geht.
5.	Im „geschützten Desktop“ (STRG+ALT+Entf) erfolgt keine Ausgabe mehr auf der Braillezeile, weil jene auf Benutzer-Ebene ja gekoppelt ist. Bei NVDA 2017.4 war dem nicht so. Hier wurde der Braillezeilen-Inhalt in beiden Umgebungen automatisch umgeschaltet. bei NVDA 2018.1 bleibt stattdessen „Geschützter Desktop“ auf der Braillezeile stehen, sobald STRG+ALT+Entf gedrückt wurde und die Braillezeile zuvor auf Benutzer-Ebene gekoppelt war.
6.	Unter NVDA 2017.4 war beim Versetzen des PCs in den Standby-Modus kurz noch ein blinkender Cursor (Punkte 7 und 8) wegen dem Kennwort-Eingabe-Textfeld auf der ALVA BC680 zu sehen. Bei NVDA 2018.1 ist dem nicht mehr so. (Hängt mit Punkt 5 zusammen.)
7.	Schaltet man die Braillezeile zuerst ein und weckt erst danach den PC aus dem Standby-Modus wieder auf, so blinkte der Cursor unter NVDA 2017.4 nach wenigen Sekunden im Kennwort-Eingabe-textfeld wieder weiter. Bei NVDA 2018.1 ist dem nicht mehr so. (Hängt mit Punkt 5 zusammen.)
8.	Schaltet man die Verbindungsarten zwischen USB (damit ist USB1 gemeint), USB2 und BT (damit ist Bluetooth gemeint) mit den rechten Smartpad-Tasten SP2+SP3+SPUp bzw. SP2+SP3+SPDown um, so reagiert die Braillezeile nach dem zurückschalten auf USB2 unter NVDA 2018.1 gar nicht mehr. Es wird weder neuer Text auf der Braillezeile ausgegeben, noch funktionieren die Tasten auf der Braillezeile. Die Braillezeile muss „doppelt“ neu gekoppelt werden, damit jene wieder benutzbar wird. Es spielt hierbei auch keine Rolle, ob via USB1 oder via Bluetooth zu einem anderen Gerät tatsächlich auch eine Verbindung hergestellt worden ist, bevor wieder auf „USB2“ zurückgeschalten wird. Via USB1 tritt dieser Fehler unter NVDA 2018.1 nicht auf. Unter NVDA 2017.4 tritt dieser Fehler weder bei USB1 noch bei USB2 auf.
9.	Durch das gleichzeitige Betätigen von SP1+SP2+SP3, SP2+SP3+Sp4 und SP1+SP2+SP3+SP4 bleibt unter NVDA 2018.1 sowohl via USB1 als auch via USB2 jeweils mindestens einer dieser vier Smartpad-Tasten „hängen“. Betätigt man gleich daraufhin z.B. „etouch1“, so wird von NVDA „SP1+Sp2+Sp3+etouch1“ erkannt. Durch mehrfaches, einzelne Drücken der vier zuvor genannten Smartpad-Tasten kann dieses „Hängenbleiben“ manchmal wieder gelöst werden. Alternativ muss halt die Braillezeile „doppelt“ neu gekoppelt werden, damit wieder alle Bedientasten auf der ALVA BC680 ordnungsgemäß funktionieren. Unter NVDA 2017.4 trat dieses Problem ebenfalls noch nicht auf.
10.	[Beschreibt einen anderen Bug, der bereits in NVDA 2015.3 existiert. Deshalb steht er hier nicht.]

### LINKED ISSUE #8149: Combining emulated keys via a braille display ignores "gestures.ini"
NVDA Version: nvda_snapshot_try-i8106-14994,fcd1ac68
Run as portable version. (It doesn't contain my huge "gestures.ini" file from my installed NVDA version. Well, it doesn't contain anything from the installed version at all.)
Braille display: Optelec ALVA BC680 (firmware: 3.0.2, connected via USB2 to the PC)

Default key mapping ("gestures.ini" doesn't exist):
Current Result: SP2 = ALT; SPLeft = Left Arrow; SP2+SPLeft = ALT+Left Arrow

Then open the input gestures window and delete the SP2 entry allocated with the ALT key (emulated hotkeys).

The "gestures.ini" now exists and contains the following line: "None = br(alva):sp2" (line 2).

Current Result: SP2 = [none]; SPLeft = Left Arrow; SP2+SPLeft = ALT+Left Arrow
Expected Result: SP2 = [none]; SPLeft = Left Arrow; SP2+SPLeft = [none]

then open the input gestures window again an allocate BC680-SP2 with the command to say the current time and date.

The following line was added in the "gestures.ini" below line 2: "dateTime = br(alva.bc680):sp2"

Current Result: SP2 = dateTime; SPLeft = Left Arrow; SP2+SPLeft = ALT+Left Arrow
Expected Result: SP2 = dateTime; SPLeft = Left Arrow; SP2+SPLeft = [none]

Then delete the "dateTime" via the input gesture window and allocate this command with SP2. The 2nd line with the "None" entry in the "gestures.ini" was deleted now.

The 2nd line contains the following: "dateTime = br(alva):sp2". The 3rd line was of course deleted too.

Current Result: SP2 = dateTime; SPLeft = Left Arrow; SP2+SPLeft = ALT+Left Arrow
Expected Result: SP2 = dateTime; SPLeft = Left Arrow; SP2+SPLeft = [none]

Then open the "gestures.ini" directly in Notepad++ (or any other source code editor) and add the following line followed by an empty line: "None = br(alva):sp2+spleft". After saving this file restart NVDA.

Current Result: SP2 = dateTime; SPLeft = Left Arrow; SP2+SPLeft = ALT+Left Arrow
Expected Result: SP2 = dateTime; SPLeft = Left Arrow; SP2+SPLeft = [none]

And the last step: Open the input gestures window and add SP2+SPLeft to the "dateTime" entry. The "None" line in the "gestures.ini" was deleted after that. The 2nd line contains the following: "dateTime = br(alva):sp2+spleft, br(alva):sp2".

Current Result: SP2 = dateTime; SPLeft = Left Arrow; SP2+SPLeft = dateTime
Expected Result: SP2 = dateTime; SPLeft = Left Arrow; SP2+SPLeft = dateTime

This new bug in NVDA 2018.1 might be the accidently result of #7783. It can be solved if NVDA considers the "gestures.ini" as well as checks if all pressed keys on the braille display are allocated with an emulated key. If only one of those pressed keys on the braille display is not allocated with an emulated key NVDA shouldn't combine emulated keys.

I just now tested the whole thing with two keys on my ALVA BC680 via the USB2 port and within changing the allocations of only two commands to keep it simple. There are thousands of combinations which can be pressed alone with the ALVA BC680. And by the way I don't think that this is an ALVA specific issue. For me it looks more like a general issue in this new function.

### LINKED ISSUE #8360: ALVA: SP2 isn't allocated with the emulated ALT key
NVDA-Version: 2018.2 RC2, Portable, Win7x64

The emulated ALT key isn't allocated with SP2 on a ALVA BC680. This bug should be easy to fix. (I know that all the other bugs (PR #8230, #8240 and #8245) aren't fixed in 2018.2 RC2. I guess that this will be done in 2018.3.)

Quote from PR #8230:
> File: brailleDisplayDrivers/alva.py
> • Line 417: "kb:alt": ("br(alva):sp2",),
> • Line 435: "kb:alt": ("br(alva):alt",),
> This maybe could fail. Please change it to the following:
> • Line 417: "kb:alt": ("br(alva):sp2","br(alva):alt",),
> • Delete line 435.

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixes #8106 
Fixes #8074 
Fixes #8360

### Summary of the issue:
* Alva Satellite has 2 Satellite key pads, one on the leftside of display and one on the rightside with 6 keys on each key pad, but NVDA treats them as one key pad. This also applies to the BC680 smart pad
* Some key combinations on ALVA displays trigger internal functionality (e.g. they open the menu). These keys ought to be ignored by NVDA.
* In some cases, NVDA doesn't recognize a connected BC680 device
* Using combined emulated keyboard gestures results into an error sound when one of the emulated gestures is removed (i.e. set to None in gestures.ini).

### Description of how this pull request fixes the issue:
* The driver now differentiates between the two groups of smart pad and etouch keys at the model level, while maintaining backwards compatibility with old gesture assignments. For example, the left sp1 on a bc680 has two identifiers:
    + br(alva):sp1
    + br(alva.bc680):lsp1
    This will also provide a fix for #8074 
* Gestures that activate internal functions are now forcefully bound to None
* When initializing the driver for an HID display, an extra check has been implemented to make sure that the proper number of cells is used for a device.
* When checking for combined emulated gestures, make sure that we only deal with non None scripts. Note that this is only a partial fix in the sense that it does not fix #8149. I'll file a separate pr for that issue. I propose to leave this fix out of the bugfixes section for now, as it might cause wrong expectations.

### Testing performed:
Several tests by @DrSooom as noted in #8106

### Known issues with pull request:
None

### Change log entry:
* Changes
    + For ALVA BC680 and protocol converter displays, it is now possible to assign different functions to the left and right smart pad, thumb and etouch keys.
    + For ALVA BC6 displays, the key combination sp2+sp3 will now announce the current date and time, whereas sp1+sp2 emulates the Windows key. (#8230)

* Bug fixes
    + ALVA BC680 braille displays no longer intermittently fail to initialize. (#8106)
    + By default, ALVA BC6 displays will no longer execute emulated system keyboard keys when pressing key combinations involving sp2+sp3 to trigger internal functionality. (#8230)
    + Pressing sp2 on an ALVA BC6 display to emulate the alt key now works as advertised. (#8360)

### CODE REVIEW

**Comment by leonardder** [comment]
@DrSooom: just before I ask for review by @michaeldcurran, could you please check whether [this](https://ci.appveyor.com/api/buildjobs/214yc9kd4v243vt3/artifacts/output%2Fnvda_snapshot_try-i8106-15090%2C6ccc39da.exe) works as advertised in the pr description? Note that this *does not* contain fixes for #8149.

**Comment by DrSooom** [comment]
File: brailleDisplayDrivers/alva.py
* Line 417: `"kb:alt": ("br(alva):sp2",),`
* Line 435: `"kb:alt": ("br(alva):alt",),`

This maybe could fail. Please change it to the following:
* Line 417: `"kb:alt": ("br(alva):sp2","br(alva):alt",),`
* Delete line 435.

**Comment by DrSooom** [comment]
1. Well, firstly I was right regarding to the emulated ALT key - it isn't allocated to SP2 by default.
2. In the Input Help Mode NVDA still speaks "t1" for "lt1" and "rt1" on the ALVA BC680.
3. "br(alva.bc680):sp1" is shown in the Input Gestures window but now have no effects. That means that all commands has to be allocated to the new l- and r-keys.
4. Is it possible to implement a replacement command for the "gestures.ini" after the first start of NVDA (installed and portable), so all non-l/r-braille-keys will automatically be changed from "br(alva.bc680):t*" to "br(alva.bc680):rt*". … Ah, no, forget this, because there are too many combinations so the replacement routine maybe will just destroy the whole "gestures.ini". All users of NVDA 2018.1 and 2018.1.1 which have allocated commands to "br(alva.bc680):*" will have to allocate all again.
5. Or is it possible that NVDA interprets "br(alva.bc680):t1" by default as "br(alva.bc680):lt1". Then all allocations which were done before will still work in 2018.2 and higher but only on the left section on the ALVA BC680. If the user also wanted that it works also on the right section he/she has to allocated it again. That might be a better solution.
6. Well, and in the end initialling the ALVA BC680 worked correctly without any issues. (Braille split mode was not tested.)

**Comment by DrSooom** [comment]
7. Quick test: Correct recognition of the 15 etouch combos and the 15 SmartPad combos in the Input Help Mode. I didn't test all 31 Thumb combos but most of them should work as well as expected. Well, and pressing SP1+SP2+SP3+SP4 on the left and on the right section at the same time is also correctly recognized by NVDA - all 8 keys are spoken. Testing this was a little bit tricky. ;)

**Comment by leonardder** [comment]
[**@DrSooom**](https://github.com/DrSooom) commented on [3 May 2018, 18:48 CEST](https://github.com/nvaccess/nvda/pull/8230#issuecomment-386360954 "2018-05-03T16:48:08Z - Replied by Github Reply Comments"):
> 4. Is it possible to implement a replacement command for the "gestures.ini" after the first start of NVDA (installed and portable), so all non-l/r-braille-keys will automatically be changed from "br(alva.bc680):t*" to "br(alva.bc680):rt*". … Ah, no, forget this, because there are too many combinations so the replacement routine maybe will just destroy the whole "gestures.ini". All users of NVDA 2018.1 and 2018.1.1 which have allocated commands to "br(alva.bc680):*" will have to allocate all again.
> 5. Or is it possible that NVDA interprets "br(alva.bc680):t1" by default as "br(alva.bc680):lt1". Then all allocations which were done before will still work in 2018.2 and higher but only on the left section on the ALVA BC680. If the user also wanted that it works also on the right section he/she has to allocated it again. That might be a better solution.

This is really a bit out of scope for this issue. Note that you should actually only create bindings for model specific gestures if you have multiple models and you really believe that assignments between models should differ. For example, Handy Tech and Hims devices have some key assignments that differ across models by default.

The general advise is to assign gestures to non model specific identifiers, i.e. the ones that start with alva: not the ones starting with alva:bc680.

**Comment by DrSooom** [comment]
@leonardder: As long as bug #8108 isn't solved it isn't possible to see in the Input Gestures window if a command is allocated with the obsolete br(alvabc6)- or with the new br(alva)-keys. So at the moment the only solution is to set up device specific hotkeys to see a different in the Input Gestures window. And no, deleting the complete "gestures.ini" is no solution. ;)

8. Because I forgot this yesterday: The five non-allocated combos don't work as expected – it still combining Win and UpArrow for SP2+SP3+SPUp. `None = br(alva.bc680):lsp2+lsp3+lspUp` in the "gestures.ini" also didn't helped as well as `None = br(alva):sp2+sp3+spUp`. But `None = br(alva):sp2+sp3` worked as expected - disables the emulating WIN key correctly. Seems we have to null those seven braille hotkeys completely.
Hmm… `dateTime = br(alva):sp2+sp3+spup` is working correctly. And `None = br(alva):sp2+sp3+spup`? No. "spup" and "spUp" makes no different. Well, we could also say `kb:control = br(alva):sp2+sp3+spUp`. This works fine and make in the most cases no damages.
And sadly I just now recognized that you forgot the quotes in line 327 before and after the word "None". Maybe the quotes could solve this issue. Otherwise we just have to null it or allocate it with the emulated CTRL key as a workaround.
9. And "br(alva):sp2+sp2" is now a valid value.
* `dateTime = br(alva):sp2`
The left and the right SP2 has the same command, but both SP2 at the same time has no command. It doesn't cares if the left or right SP2 was pressed in the Input Gestures window. this worked as expected.
* `dateTime = br(alva):sp2` and `say_battery_status = br(alva):sp2+sp2`
The left and the right SP2 still has the same command, but both SP2 at the same time has now the other command. this also worked as expected.

**Comment by leonardder** [comment]
[**@DrSooom**](https://github.com/DrSooom) commented on [4 mei 2018 15:44 CEST](https://github.com/nvaccess/nvda/pull/8230#issuecomment-386605595 "2018-05-04T13:44:08Z - Replied by Github Reply Comments"):
> And no, deleting the complete "gestures.ini" is no solution. ;)

I'm afraid that at some point, there isn't much else for you to do other than that, since a gestures.ini can get quite dirty when you add and remove gestures for different driver many times.

> And sadly I just now recognized that you forgot the quotes in line 327 before and after the word "None". Maybe the quotes could solve this issue.

This is intentional. "None" just translates to None in inputCore.py at line 282.

Looking at the code again, I see why assigning all these combinations to None doesn't make sense, because NVDA doesn't differenciate between forcefully disabled gestures and gestures that are just not in use.

I guess in this case, it is the safest to take away the sp2+sp3 assignment to leftWindows. When we do that, sp2+sp3 translates to alt+escape, and sp2+sp3+whatever else won't be in use.

Note that there are more drivers suffering from the issue that some keys are processed by NVDA while activating internal device functionality.

@dkager: What do you think about removing the sp2+sp3 to left windows assignment altogether?

**Comment by DrSooom** [comment]
@leonardder: You can also change the alva.py as follows:
* line 425: `"dateTime": ("br(alva):sp1+sp2",),`
To: `"dateTime": ("br(alva):sp2+sp3",),`
* line 429: `"kb:windows": ("br(alva):sp2+sp3","br(alva):windows",),`
To: `"kb:windows": ("br(alva):sp1+sp2","br(alva):windows",),`

File: source/brailleDisplayDrivers/alva.py [hunk 1]
```diff
@@ -429,27 +451,39 @@ class InputGesture(braille.BrailleDisplayGesture, brailleInput.BrailleInputGestu
...
+					keyName = ALVA_KEYS[group][number]
 				except (KeyError, IndexError):
 					log.debugWarning("Unknown key with group %d and number %d" % (group, number))
-
+					return
+				names.append(keyName)
+				if isNoBC640:
+					doubledKeyCount = DOUBLED_KEY_COUNTS.get(group)
+					if doubledKeyCount:
+						keyName = ("r" if number>=doubledKeyCount else "l") + keyName
```
Comment (dkager) [inline source/brailleDisplayDrivers/alva.py]: Please add spaces around `>=`.

File: source/brailleDisplayDrivers/alva.py [hunk 2]
```diff
@@ -429,27 +451,39 @@ class InputGesture(braille.BrailleDisplayGesture, brailleInput.BrailleInputGestu
 
 	def __init__(self, model, keys, brailleInput=False):
 		super(InputGesture, self).__init__()
+		isNoBC640 = model != ALVA_MODEL_IDS[0x40]
```
Comment (dkager) [inline source/brailleDisplayDrivers/alva.py]: Can you use a constant in place of 0x40?

**Review by dkager — CHANGES_REQUESTED** [review]

**Review by dkager — DISMISSED** [review]

**Review by michaelDCurran — DISMISSED** [review]

**Review by michaelDCurran — APPROVED** [review]

**Review by michaelDCurran — APPROVED** [review]

**Pull request merged**